### PR TITLE
fix(ingest/iceberg): update iceberg source to support newer versions of pyiceberg at runtime

### PR DESCRIPTION
**Pull request by cccs-eric**
## Goal
The goal of this PR is to let users manually install newer versions of pyiceberg at runtime, which will make ingesting tables from SQL catalog possible (SQLCatalog was introduced in pyiceberg 0.5.0. 

## How
There are dependency conflicts that occur when trying to upgrade the pyiceberg package dependency to something newer than 0.4.0.  The biggest one is pydantic v2 (since pyiceberg 0.5.0) and there is another possible one with sqlalchemy v2 (depending on which extras you are pulling from pyiceberg).  Not much has changed in the pyiceberg API since 0.4.0 and it is possible to make the Iceberg source compatible with newer version of pyiceberg. 

Once this PR is merged, it will be possible for a user to ingest Iceberg tables with a newer version of pyiceberg (like 0.6.1) using the following flow:
```bash
# Notice we are not installing the iceberg extra here since it will pull pyiceberg==0.4.0
pip install acryl-datahub
# Install the latest pyiceberg version and the required extras.  For example, ingesting from a SQLCatalog in Azure would require
# pip install pyiceberg[sql-postgres,adlfs]
pip install pyiceberg[<required extras>]

datahub ingest run --config <path to your recipe.yaml>
```
where your recipe could look like this:
```yaml
source:
  type: "iceberg"
  config:
    env: PROD
    catalog:
      # REST catalog configuration example using S3 storage
      my_rest_catalog:
        type: rest
        # Catalog configuration follows pyiceberg's documentation (https://py.iceberg.apache.org/configuration)
        uri: http://localhost:8181
        s3.access-key-id: admin
        s3.secret-access-key: password
        s3.region: us-east-1
        warehouse: s3a://warehouse/wh/
        s3.endpoint: http://localhost:9000
      # SQL catalog configuration example using Azure datalake storage and a PostgreSQL database
      # my_sql_catalog:
      #   type: sql
      #   uri: postgresql+psycopg2://user:password@sqldatabase.postgres.database.azure.com:5432/icebergcatalog
      #   adlfs.tenant-id: <Azure tenant ID>
      #   adlfs.account-name: <Azure storage account name>
      #   adlfs.client-id: <Azure Client/Application ID>
      #   adlfs.client-secret: <Azure Client Secret>
    platform_instance: my_rest_catalog
    table_pattern:
      allow:
        - marketing.*
    profiling:
      enabled: true

sink:
  # sink configs
```

## Extra contributions

- This PR also changes how to configure the pyiceberg catalog inside a recipe.  The format used by the Datahub source configuration is now the same as what you would find in your `.pyiceberg.yaml`, which makes configuration of your recipe easier.
-  This PR adds a new dataset property named `partition-spec` that will expose the Iceberg table partitioning spec.  An example of such a property value could be: `[{\"name\": \"timeperiod_loaded\", \"transform\": \"identity\", \"source\": \"timeperiod_loaded\", \"source-id\": 19, \"source-type\": \"date\", \"field-id\": 1000}]`.  The structure of that JSON follows this following [spec from Iceberg](https://iceberg.apache.org/spec/?#partition-specs).

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Review by hsheth2 — COMMENTED** [review]
We can also make some changes to setup.py
- Can also change `"pyiceberg~=0.4",` -> to `"pyiceberg>=0.4,<0.7",` and remove the pydantic_no_v2 requirement
- Can add iceberg to the exception list around line 798

That way iceberg 0.6 can be installed natively

File: metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg_common.py
Comment (cccs-eric) [reply metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg_common.py:85]: No, it is not.  If not provided, pyiceberg will infer its value from `uri`.
https://github.com/apache/iceberg-python/blob/main/pyiceberg/catalog/__init__.py#L219-L223

You could argue that `uri` is required, but I did not want to pull all the pyiceberg config validation into Datahub and leave it to pyiceberg.  If a value is missing, pyiceberg will raise an error.

File: metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg.py
Comment (cccs-eric) [reply metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg.py:237]: Odds are small, but you are right about this.  A broken table could trigger an error, and an error at this stage should not prevent ingestion of the table.  I'll add guards and a warning in the logs if an error occurs.

File: metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg_common.py
Comment (cccs-eric) [reply metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg_common.py:59]: I thought that since the iceberg source was still consider in `testing`, it was ok to introduce a breaking change.  I'll see what I can do.

Comment (cccs-eric) [reply metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg_common.py]: I now support the old format.  One thing that is left is to report the warning back to the user.  What would be your recommendation?  It currently `print()` a message, but I will change this to what ever you recommend.

**Review by hsheth2 — APPROVED** [review]

File: metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg_common.py
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg_common.py:63]: actually I think the type here should be only `Dict[str, Dict[str, str]]` as you had previously

the validator will automatically convert the old format to the new one, but we don't want to show the old type in the docs, and we want the correct type annotation so that mypy can check our usages of the configs

File: metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg_common.py
Comment (cccs-eric) [reply metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg_common.py]: @hsheth2 Thanks for approving it, but don't merge it just yet.  The warning back to the user needs to be fixed.

File: metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg_common.py
Comment (cccs-eric) [reply metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg_common.py:63]: You are absolutely right!  I made that change before implementing the validator.

File: metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg_common.py
Comment (cccs-eric) [reply metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg_common.py:62]: I ended up simply using `logging.warning()`.  Do we have to state anywhere in the docs that the old format is deprecated?

Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg_common.py:62]: logger.warning should be fine here

**Review by hsheth2 — APPROVED** [review]

**Comment by usmanovbf** [comment]
Good news, thank you! Can't wait to see it in master

**Pull request merged**
